src/classes/RD2_RecurringDonations_TDTM.cls

```diff
@@ -252,124 +251,180 @@ public class RD2_RecurringDonations_TDTM extends TDTM_Runnable {
+                    contactIds.add(rd.npe03__Contact__c);
+                }
+            }
+            if (contactIds.isEmpty()) {
+                return;
+            }
+
+            Map<Id, Contact> contactById = new Map<Id, Contact>([
+                SELECT Name, AccountId
+                FROM Contact
+                WHERE Id IN :contactIds
+            ]);
+
+            for (npe03__Recurring_Donation__c rd : rds) {
+                if (rd.npe03__Organization__c == null
+                    && rd.npe03__Contact__c != null
+                    && contactById.containsKey(rd.npe03__Contact__c)
+                ) {
+                    rd.npe03__Organization__c = contactById.get(rd.npe03__Contact__c).AccountId;
+                }
+            }
+        }
+
+        /***
+        * @description Sets schedule fields
+        * @param rd Recurring Donation record
+        * @return void
+        */
+        private void adjustScheduleFields(npe03__Recurring_Donation__c rd) {
+            if (rd.npe03__Installment_Period__c != RD2_Constants.INSTALLMENT_PERIOD_MONTHLY && rd.Day_of_Month__c != null) {
+                rd.Day_of_Month__c = null;
+            }
+
+            //map the Day of Month value of "31" to the Last_Day value
+            if (rd.Day_of_Month__c == '31') {
+                rd.Day_of_Month__c = RD2_Constants.DAY_OF_MONTH_LAST_DAY;
+            }
 
-            if (newRd.isClosed() && rds[i].EndDate__c == null) {
-                rds[i].EndDate__c = currentDate.addDays(-1);
+            //set the default installment frequency when value is not provided on the RD insert
+            if (rd.InstallmentFrequency__c == null && rd.Id == null) {// Record Id is not available in the "before insert" context
+                rd.InstallmentFrequency__c = RD2_Constants.DEFAULT_INSTALLMENT_FREQUENCY;
+            }
+        }
+        
+        /***
+        * @description Evaluate End Date on a Recurring Donation record
+        * If the RD is just Closed and the End Date is not provided, auto-populate the End Date to yesterday.
+        * If the RD is just Closed but the End Date is in the future, revert the status to the value as it was before the update.
+        * If the RD is reopened, clear out the End Date.
+        * @param rd Recurring Donation record
+        * @param oldRd Old version of the Recurring Donation record
+        */
+        private void adjustEndDateAndStatus(npe03__Recurring_Donation__c rd, npe03__Recurring_Donation__c oldRd) {
+            RD2_RecurringDonation rdRecord = new RD2_RecurringDonation(rd);
+
+            if (rdRecord.isClosed() && rd.EndDate__c == null) {
+                rd.EndDate__c = currentDate.addDays(-1);
             }
 
-            if (triggerAction == TDTM_Runnable.Action.BeforeUpdate) {
-                if (!newRd.isClosed() && (new RD2_RecurringDonation(oldRds[i]).isClosed())) {
-                    rds[i].EndDate__c = null;
+            if (oldRd != null) {
```

Comment:
Checking for `oldRd` or record Id in order to differentiate between update and insert so checks are applicable when validation is called from a trigger or the entry form (W-8247722).